### ISSUE TITLE
Documentation is confusing for request generators

### ISSUE THREAD

**Issue by iainbryson** [alexfernandez/loadtest#86]
In this section:

```javascript
module.exports = function(params, options, client, callback) {
  generateMessageAsync(function(message) {

    if (message)
    {
      options.headers['Content-Length'] = message.length;
      options.headers['Content-Type'] = 'application/x-www-form-urlencoded';
    }
    request = client(options, callback);
    if (message){
      request.write(message);
    }

    return request;
  }
}
```

It's unclear what "generateMessageAsync" is.  There's no such function in the project and the syntax is funky.  I suppose this should be replaced by something like

```
var message = { your: "message here" };
```

Or a function to generate the message _synchronously_.  It's not clear how the message could ever be generated asynchronously since request isn't a function which is returned to be called.

I can fix this if you assign this to me.

**Comment by alexfernandez** [comment]
Please study the code carefully. The generator is actually asynchronous, and it generates a client that contains the callback.

I made you a collaborator so I can assign you to the issue to study it. Please continue sending pull requests, I will do the same with any significant changes, OK?

**Comment by iainbryson** [comment]
Alex, let me try to explain this from a different angle.  In the example `request-generator.js`, the request generator is a simple function taking the 4 parameters and returning the client with the embedded callback:

```
			requestGenerator: function(params, options, client, callback)
			{
				var message = '{"hi": "ho"}';
				options.headers['Content-Length'] = message.length;
				options.headers['Content-Type'] = 'application/json';
				var request = client(options, callback);
				request.write(message);
				return request;
			},
```

However, in the README section above, request generator has this strange getMessageAsync line in it, which appears nowhere in the source tree other than the README.  And the 4-parameter function in the README doesn't actually return anything -- it just calls this getMessageAsync function with the anonymous function. 

When I say it's called synchronously, this is what I mean: In `loadtest.js`, the request generator is called and the return value is used subsequently.  I supposed this is just a terminology confusion.

```
		if (typeof params.requestGenerator == 'function') {
			request = params.requestGenerator(params, options, lib.request, getConnect(id, requestFinished));
		} else {
			request = lib.request(options, getConnect(id, requestFinished));
		}

		if (params.hasOwnProperty("timeout")) {
			request.setTimeout(params.timeout, function() {
				requestFinished('Connection timed out');
			});
		}

		if(message)
		{
			request.write(message);
		}

		request.on('error', function(error)
		{
			requestFinished('Connection error: ' + error.message);
		});
		request.end();
```

(Aside: It's a little odd that you check for the message even if the request generator is used.)

So, bottom line is that I think the `README.md` examples should reflect the `request-generator.js` code and simply do something like:

```
module.exports = function(params, options, client, callback) {
   var message = "message";

      options.headers['Content-Length'] = message.length;
      options.headers['Content-Type'] = 'application/x-www-form-urlencoded';

    request = client(options, callback);

      request.write(message);

    return request;
}
```

That's when I ended up doing in my own module, which I why I brought this up ;)

**Comment by alexfernandez** [comment]
OK, since you are clear about it, feel free to fix it! :)

**Comment by jjohnsonvng** [comment]
Can the Request Generator generate custom Body and URL?

What all properties are available on the options parameter.

**Comment by alexfernandez** [comment]
Please see (`lib/requestGenerator.js`](https://github.com/alexfernandez/loadtest/blob/master/lib/request-generator.js) for an example which sends a custom body: `message`. As to the URL, you have full control of the `options` object, which is passed directly to [`http.request()`](https://nodejs.org/api/http.html#http_http_request_options_callback), or `https.request` for secure requests.

The `params` object has [all these attributes](https://github.com/alexfernandez/loadtest/blob/master/lib/loadtest.js#L316).

**Comment by jjohnsonvng** [comment]
I think I am missing the point with this.  I was thinking this would be a great way to run a Large amount of tests that would randomly switch between various calls.

**Comment by alexfernandez** [comment]
Sure, why not? You can customize the body and the URL, then create your own `options` object to pass to `http.request()`. Once the request [has been used](https://github.com/alexfernandez/loadtest/blob/master/lib/request-generator.js#L34) just return it:

```
				var request = client(customOptions, callback);
				request.write(customMessage);
				return request;
```

and loadtest will do the rest for you (measurements and display of results).

**Comment by jjohnsonvng** [comment]
Do you have an example of a LoadTest doing this?

**Comment by alexfernandez** [comment]
You can use the linked example as a start. I don't have anything else, sorry.

**Comment by jjohnsonvng** [comment]
ok so in that case in the requestGenerator I can change method, body and url before I var request = client(options, callback); 

Is that correct?

**Comment by alexfernandez** [comment]
Yes, sure! To change URL you can even pass a URL to `client`, which is either [`http.request`](https://nodejs.org/api/http.html#http_http_request_options_callback) or `https.request`. Or you can use a more sophisticated `options` object.

**Comment by jjohnsonvng** [comment]
Well this all basically works except one thing.   options does not carry the body property so how can I change the body I am posting?

**Comment by alexfernandez** [comment]
I guess it's not possible, you have to generate a new body from scratch. It would not take much time to add `message` as either `params.message` or even `options.message` so you can modify it, I guess nobody has felt the need so far. Patches are welcome :D

**Comment by jjohnsonvng** [comment]
Were is the code for Client and I will look at making a mode ClientWithBody

**Comment by alexfernandez** [comment]
The `requestGenerator` is invoked from [here](https://github.com/alexfernandez/loadtest/blob/master/lib/loadtest.js#L204).

**Comment by jjohnsonvng** [comment]
so your just using the request method in http npm it looks like?

**Comment by alexfernandez** [comment]
Yes, basically.

**Comment by jjohnsonvng** [comment]
Ok here is a script that runs for anyone else
var options = {
    url: 'http://yourHost',
    concurrent: 5,
    method: 'POST',
    body:'',
    requestPerSecond:5,
    maxSeconds:30,
    requestGenerator: function(params, options, client, callback)
			{
				var message = '{"hi": "ho"}';
				options.headers['Content-Length'] = message.length;
				options.headers['Content-Type'] = 'application/json';
				options.body = 'YourPostData'
                                options.path = 'YourURLPath';

				var request = client(options, callback);
				request.write(message);
				return request;
			}
    
};

loadtest.loadTest(options, function(error, results)
{
    if (error)
    {
        return console.error('Got an error: %s', error);
    }
    console.log(results)
    console.log('Tests run successfully');
});

**Comment by alexfernandez** [comment]
Cool, thanks! Care to add your example as a sample file?

**Comment by jjohnsonvng** [comment]
Feel free to  and as I come up with more Complex Tests I will send them over

**Comment by alexfernandez** [comment]
Included in 1.4.2: https://github.com/alexfernandez/loadtest/blob/master/sample/request-generator.js
